### DISCUSSION TITLE
cannot launch the `wholebodydynamics` and `gravitycompensator` when doing the redball demo

### DISCUSSION THREAD

**Discussion by Yujia-Zhang0913** [robotology/community#661] [Q&A] [answered]
Hi, I follow the guide of tech-doc but I still cannot launch the wholebodydynamics.
The error is like the picture below. I do not know why.
![图片](https://github.com/robotology/community/assets/73634487/19719966-5867-4c68-9d5d-4eafef2797a2)
Can you help me?
Moreover, sometimes the wholebodydynamics can be launched, but the gravitycompensater cannot be launched. The inertial part cannot be connect(Since now I cannot launch the dynamics, I cannot reproduce the error of gravitycompensator. So I don't have the screenshot of this error.)
Really thanks.

**Comment by pattacini** [comment]
Hi @Yujia-Zhang0913 

From the screenshot you shared is somewhat difficult to figure out what the problem is.
Could you please drop herein the log file as exported from the `yarplogger` when the problem occurs?

Also, to help us narrow down this snag, could you tell us the following?
- What is your robot name?
- What is the SW distro your robot is currently using?

You may easily find this information by ssh-ing into `icub-head` and typing:
```console
echo $YARP_ROBOT_NAME
cd $ROBOT_CODE
cd robotology-superbuild
git status
```
and reporting here the output.

**Reply by Yujia-Zhang0913** [reply]
Hi,
![Screenshot from 2023-10-27 04-46-12](https://github.com/robotology/community/assets/73634487/176cd5df-4d14-471d-bc1b-5c3d7f4506b3)
[log_icubsrv_wholeBodyDynamics_6187.log](https://github.com/robotology/community/files/13184496/log_icubsrv_wholeBodyDynamics_6187.log)
And I am new to icub, so I am still learning how to manipulate it.
And when I want to do the yarpdataplayer, the video is played but the robot doesn't move. Could you help me with this, please. Really thanks.
![Screenshot from 2023-10-27 04-54-55](https://github.com/robotology/community/assets/73634487/c8436277-f9b4-442e-a307-c2204bd36b54)

**Reply by pattacini** [reply, marked as answer]
Hi @Yujia-Zhang0913 

The problem is that the inertial device is disabled on your robot.
See https://github.com/robotology/robots-configuration/blob/master/iCubSingapore01/icub_all.xml#L52-L56.

We made this choice because we no longer support old devices like the one mounted on the iCubSingapore01 head. The most recent iCub robots sport the RFE board for which our SW works nicely.

No worries though. You can uncomment those lines reported above to reinstate the inertial device with the caveat that the first cold start of the `yarprobotinterface` won't work. You'll need to launch `yarprobotinterface` twice in a row, indeed. A bit annoying but doable.

cc'ing @AntonioConsilvio @sgiraz for this.

To check which configuration file you need to update, do this:
- ssh into `icub-head`.
- run:
  ```console
   yarp resource --context yarprobotinterface --from icub_all.xml 
  ```
- note down which file was found.
- edit that specific file to uncomment those lines.

### Running the red ball demo without WBD
Be aware that you don't specifically need to have `wholebodydynamics` and `gravitycompesator` to be up and running in order to launch the Red-Ball-Demo. 

Thus, you could leave the `icub_all.xml` file unchanged and then tune up the red-ball-demo configuration file.

On the machine where you launch the module `demoRedBall`[^1], do:
- run:
  ```console
   yarp resource --context demoRedBall --from config.ini
  ```
- note down which file was found.
- edit that specific file to set to `off` the options `impedance_velocity_mode`:
    - https://github.com/robotology/icub-basic-demos/blob/master/demoRedBall/app/conf/config.ini#L25
    - https://github.com/robotology/icub-basic-demos/blob/master/demoRedBall/app/conf/config.ini#L34

---

>And when I want to do the yarpdataplayer, the video is played but the robot doesn't move. Could you help me with this, please. Really thanks.

That example works only in simulation and not on a physical robot.
[^1]: You can figure it out from the node reported in `yarpmanager`.

**Reply by Yujia-Zhang0913** [reply]
Hi, @pattacini 
I found that icub cannot move their hands to the accurate positions and I think it is because the torso is not in the right pose when init the robot. And when I launch the demo of red ball, it will always do grasp beneath the ball. 
So I think maybe I should do the calibration? However, there is different version of icub calibration [https://icub-tech-iit.github.io/documentation/icub_robot_calibration/icub-robot-calibration-v2.x/](url)
![Screenshot from 2023-11-08 03-30-25](https://github.com/robotology/community/assets/73634487/8a464e2b-156f-484a-9a4c-fdabbbfe71bf)
So should I follow this tutorial to calibrate? Thanks.

**Reply by sgiraz** [reply]
Hi @Yujia-Zhang0913,

`iCubSingapore01 S/N035` is equipped with the hands v2.5, and for torso calibration, you can refer to the instructions outlined in the [iCub 2.X Robot Calibration][1]

[1]: https://icub-tech-iit.github.io/documentation/icub_robot_calibration/icub-robot-calibration-v2.x

**Reply by pattacini** [reply]
To expand more on this, the red-ball demo requires an accurate calibration of all robot parts to work at its best throughout a chain of calibrations.

In detail:
- The torso, the arm, and the eyes need to be accurately calibrated following the documentation posted by @sgiraz. In particular, the arms can be further [fine-calibrated](https://icub-tech-iit.github.io/documentation/icub_robot_calibration/icub-robot-calibration-v2.x/#arms-fine-calibration). 
- The cameras also need to undergo a specific [calibration](https://icub-tech-iit.github.io/documentation/icub_robot_calibration/icub-robot-calibration-v2.x/#calibrating-cameras).
- Finally, given that the 3D estimation of the red ball is not 100% precise, you have to tinker with the [3D hand offsets](https://github.com/robotology/icub-basic-demos/blob/master/demoRedBall/app/conf/config.ini#L22-L24), which are explicitly foreseen by the demo app itself and account for spatial corrections with respect to the [hand reference frame](https://icub-tech-iit.github.io/documentation/icub_kinematics/icub-forward-kinematics/icub-forward-kinematics-arms/#left).

Most likely, you have everything spot on regarding the absolute and fine calibration of the parts, barring the slight misalignment you spotted on the torso that you can correct following the steps above[^1]. 

Most of the time, then, you'll find yourself tuning those 3D hand offsets only.

[^1]: The wear and tear tend to impact the calibration for tendon-driven mechanisms.
